File: solutions/3-Longest-Substring-Without-Repeating-Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let length = s.length
    let l = 0
    let r = 0
    let max = 0
    let map = {}
    for(;r<length&&max<length-l;r++){
        if(map[s[r]]>=l){
            if(r-l>max) max = r-l
            l = map[s[r]] + 1
        }
        map[s[r]] = r
    }
    if(r-l>max) max = r-l
    return max
}

File: solutions/3-longest-substring-without-repeating-characters.cpp
#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    int len = s.length();
    if (len < 2) {
      return len;
    }
    unordered_set<char> us;
    int max = 1;
    int i = 0, j = 0;
    while (i < len - max) {
      while (j < len && us.find(s[j]) == us.end()) {
        us.insert(s[j]);
        j++;
      }
      if (j - i > max) {
        max = j - i;
      }
      us.erase(s[i]);
      i++;
    }
    return max;
  }
};

int main() {
  Solution s;
  cout << s.lengthOfLongestSubstring("a") << endl;
  cout << s.lengthOfLongestSubstring("abcabcbb") << endl;
  cout << s.lengthOfLongestSubstring("aa") << endl;
  cout << s.lengthOfLongestSubstring("ab") << endl;
  cout << s.lengthOfLongestSubstring("pwwkew") << endl;
  cout << s.lengthOfLongestSubstring("au") << endl;
  return 0;
}